readme.md
<p align="center">
  <img src="https://i.imgur.com/l67acQe.png" height="300" alt="featherstore">  
</p>

# featherstore [![Build Status](https://travis-ci.com/ApolloStationIO/featherstore.svg?branch=master)](https://travis-ci.com/ApolloStationIO/featherstore)

> Zero-config, 600B, cross-browser storage for JavaScript apps

## Installation

`npm i featherstore --save`

or with [yarn](https://yarnpkg.com)

`yarn add featherstore`

## Usage

Creating a store a fairly easy -

```javascript
const store = featherstore();
```

### Options

You can pass in few options to `featherstore` (as an object) to create the store

- **namespace**
  Just so that you don't pollute the global namespace, the `namespace` you provide is prefixed to all the `keys` you use to save the data to the store. Defaults to `__FEATHERSTORE__`.

- **ttl** (time to live)
  That's the time in seconds for which the value will be stored. It defaults to **one day** from the time you save the value. You can pass any value that you like.

- **storage**
  The storage that'll be used to save the store data. Defaults to `window.localStorage`. You can pass in `'session'` to use `window.sessionStorage`. Else, you can supply your own storage, considering that has the same API surface as `window.localStorage`

```javascript
const store = featherstore({
  namespace: '__APPNAME__',
  ttl: 60 * 60 * 24 * 1000 * 365, // One year
  storage: 'session'
});
```

## Sample

```javascript
// create store
const store = featherstore();

// set a value
store.set('key', 'value');

// set a value with a custom timeout (other than default ttl) by passing ttl as third argument
store.set('key_3_sec', 'value for 3 secs', 3000);

// get a stored value
store.get('key'); // value

// check if the store has a value with some `key`
store.has('key'); // true
store.has('non_existing_key'); // false

// get the list of keys in store
store.keys(); // ['key']

// remove a value with some `key`
store.clear('key');

// remove all values from store i.e clear the store
store.clear();
```

## [memstore](https://github.com/ApolloStationIO/featherstore-plugin-memstore)

You can use an in-memory store (in case you would not like to use `localStorage` or `sessionStorage`). [memstore](https://github.com/ApolloStationIO/featherstore-plugin-memstore) helps you do that with just **200B** of additional weight.

## License

Copyright © MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 1.0.1 (2019-06-25)


test/setup.js
import 'babel-polyfill';
import 'jest-localstorage-mock';


test/index.spec.js
import featherstore from '../src';

let store;

beforeEach(() => {
  store = featherstore();
});

afterEach(() => {
  store.clear();
});

describe('featherstore', () => {
  it('should create a store exposing an API with five methods', () => {
    const keys = Object.keys(store);
    expect(Object.keys(store).length).toBe(5);
    expect(keys).toStrictEqual(['set', 'get', 'clear', 'keys', 'has']);
  });
});

describe('featherstore#set', () => {
  it('should save a key-value provided', () => {
    expect(store.set('key', 'value')).toBe('value');
    expect(store.get('key')).toBe('value');
  });

  it('should save key-value for a time explicitly provided', async done => {
    store.set('key', 'value for 1sec', 1000);
    await new Promise(resolve => {
      setTimeout(() => {
        resolve();
      }, 1000);
    });
    expect(store.get('key')).toBe(null);
    done();
  });
});

describe('featherstore#get', () => {
  it('should get a stored value', () => {
    store.set('number', 3);
    store.set('string', 'featherstore');
    store.set('array', [1, 2, 3]);
    store.set('object', { a: 'a' });

    expect(store.get('number')).toStrictEqual(3);
    expect(store.get('string')).toStrictEqual('featherstore');
    expect(store.get('array')).toStrictEqual([1, 2, 3]);
    expect(store.get('object')).toStrictEqual({ a: 'a' });
  });
});

describe('featherstore#keys', () => {
  it('should return all the keys stored in the store', () => {
    store.set('key1', 'value1');
    store.set('key2', 'value2');
    expect(store.keys()).toStrictEqual(['key1', 'key2']);
  });

  it('should return empty list if called on empty store', () => {
    expect(store.keys().length).toBe(0);
  });
});

describe('featherstore#clear', () => {
  it('should remove all the stored values from the storage', () => {
    store.set('number', 3);
    store.set('string', 'featherstore');
    store.set('array', [1, 2, 3]);
    store.set('object', { a: 'a' });
    store.clear();

    expect(store.keys().length).toBe(0);
  });

  it('should clear only one item (if exists) if the key is provided', () => {
    store.set('number', 3);
    store.set('string', 'featherstore');
    store.clear('number');

    expect(store.get('number')).toBe(null);
    expect(store.get('string')).toBe('featherstore');
  });
});

describe('featherstore#has', () => {
  it('should be truthy if store has a value stored with the key', () => {
    store.set('number', 3);
    expect(store.has('number')).toBe(true);
  });

  it('should be falsy if store does not have a value stored with the key', () => {
    expect(store.has('number')).toBe(false);
  });
});


src/index.js
/** Featherstore: Zero-config, 500B, cross-browser storage for JavaScript apps
 *  @name featherstore
 *  @param { object } options An object with properties `namespace`, `storage` and `ttl` (time to live)
 *  @returns { Featherstore } Returns a store with methods `get`, `set`, `has`, `keys` and `clear`
 */
export default function featherstore(options = {}) {
  const defaultOpts = {
    namespace: '__FEATHERSTORE__',
    storage: window.localStorage,
    ttl: 60 * 60 * 24 * 1000 // One day
  };

  // Choosing storage to be used
  let storage =
    options.storage === 'session' ? window.sessionStorage : window.localStorage;
  if (typeof options.storage === 'object') {
    storage = options.storage;
  }

  // Options to be used by library
  const $opts = (Object.freeze || Object)(
    Object.assign({}, defaultOpts, options, {
      storage
    })
  );

  // Utils
  const getPrefixedKey = function(key) {
    return `${$opts.namespace}${key}`;
  };

  const getUnprefixedKey = function(key) {
    return key.replace(new RegExp(`^${$opts.namespace}`), '');
  };

  const stringifyJSON = JSON.stringify;
  const parseJSON = JSON.parse;

  // Return the store
  return {
    set(key, data, ttl = $opts.ttl) {
      const prefixedKey = getPrefixedKey(key);
      const value = {
        ttl: Date.now() + ttl,
        data
      };
      $opts.storage.setItem(prefixedKey, stringifyJSON(value));
      return data;
    },

    get(key, fallback = null) {
      try {
        const prefixedKey = getPrefixedKey(key);
        const value = parseJSON($opts.storage.getItem(prefixedKey));

        if (value.data !== null) {
          if ('ttl' in value && Number(value.ttl) < Date.now()) {
            $opts.storage.removeItem(prefixedKey);
            return null;
          }

          if ('data' in value) {
            return value.data;
          }

          return fallback;
        }
      } catch (error) {
        if (process.env.NODE_ENV !== 'production') {
          console.error(
            `Featherstore [get]: Can't find any value for '${key}' in the store.`
          );
          return fallback;
        }
      }
    },

    clear(key = null) {
      if (key === null) {
        $opts.storage.clear();
      } else {
        $opts.storage.removeItem(getPrefixedKey(key));
      }

      return true;
    },

    keys() {
      return Object.keys($opts.storage).map(getUnprefixedKey);
    },

    has(key) {
      const prefixedKey = getPrefixedKey(key);
      return prefixedKey in $opts.storage;
    }
  };
}
